### PR TITLE
Use ActiveSupport::Concurrency::LoadInterlockAwareMonitor to avoid deadlock

### PR DESCRIPTION
**Pull request by sergiogomez**
### Summary

There is a deadlock between `safe_constantize` and `synchronize`, reported at https://github.com/rails/rails/issues/32451. Then, I think we can avoid that problem using `synchronize` only at the block after `safe_constantize`.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Can you please point to the master branch? We will backport to the supported releases after we fix the problem in the latest.

File: actionpack/lib/action_controller/metal/params_wrapper.rb [hunk 1]
```diff
@@ -91,7 +91,7 @@ def initialize(name, format, include, exclude, klass, model) # :nodoc:
       end
 
       def model
-        super || synchronize { super || self.model = _default_wrap_model }
+        super || self.model = _default_wrap_model
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/params_wrapper.rb]: I think what is important to synchronize is the assignment to `self.model`. There is nothing in the `_default_wrap_model` what seems to require lock.

Comment (sergiogomez) [reply actionpack/lib/action_controller/metal/params_wrapper.rb]: True, and in that case my patch is invalid. The deadlock is here: https://github.com/sergiogomez/rails/blob/af7f7c9beff52aceb2f6ab7368e50e6d7820d411/actionpack/lib/action_controller/metal/params_wrapper.rb#L157

Comment (rafaelfranca) [reply actionpack/lib/action_controller/metal/params_wrapper.rb]: Maybe what we need to do is make sure the mutex we are using is loading aware. We could try to use this class. https://github.com/rails/rails/blob/master/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb

**Comment by sergiogomez** [comment]
After commit https://github.com/rails/rails/pull/32453/commits/5fd22435a57a1633e90100d109f325ab593d7684, what do you think about this other approach @rafaelfranca ?

File: actionpack/lib/action_controller/metal/params_wrapper.rb [hunk 2]
```diff
@@ -93,7 +93,8 @@ def initialize(name, format, include, exclude, klass, model) # :nodoc:
       end
 
       def model
-        super || synchronize { super || self.model = _default_wrap_model }
+        m = _default_wrap_model
```
[the remark this answers is not in the document]
Comment (sergiogomez) [reply actionpack/lib/action_controller/metal/params_wrapper.rb]: 👍

**Comment by rafaelfranca** [comment]
This looks good to me, but I'm waiting @matthewd's review too.

**Comment by matthewd** [comment]
I'm a bit confused about which thing the synchronize was originally protecting... I'd assumed it was the assignment, but on further thought, I don't see how that would need it.

Everything else being equal, I'd rather stick to the simple lock and just accept the chance of doing wasted work, in favour of not holding any locks.

But without understanding why the lock is there in the first place, I don't feel confident pushing that as the right solution.

Relatedly, I'd want to understand the other `synchronize` calls in this class -- beyond being confusing, does it matter that this PR splits these operations across two locks? Should we drop the `Mutex_m` and move them all to use the new `@monitor`?

**Comment by sergiogomez** [comment]
Thanks @matthewd . I've updated this PR dropping the `Mutex_m` and move the other locks to the new `@monitor`.

Also, I've created [another PR](https://github.com/rails/rails/pull/32541) removing `mutex_m` lock from method `model`. What do you think about it? cc/ @rafaelfranca

**Comment by mathiasAichinger** [comment]
Looking forward to the fix. :)

**Comment by sergiogomez** [comment]
Closed in favour of https://github.com/rails/rails/pull/32541

**Pull request closed without merging**